### DISCUSSION TITLE
How can I change MaxBufferSize using Setting?

### DISCUSSION THREAD

**Discussion by RobertoHE** [lathoub/Arduino-BLE-MIDI#58] [Q&A]
Hi everyone.

I have a Roland AX-Edge device that only alow receives SysEx messages split into 2 or more packets. Because of that, if I reduce the buffer size I can send multiple packets for each SysEx correctly. I have tested that and it works fine.

I want to reduce `MaxBufferSize` to 16 but a don't want to modify it in `BLEMIDI_Settings.h` lib file, I prefer to create a local setting in my code and don't edit any source file from the library.
https://github.com/lathoub/Arduino-BLE-MIDI/issues/14#issuecomment-619626138
https://github.com/lathoub/Arduino-BLE-MIDI/blob/8dee1f7d257aa71aeb0e956632da2dd2ae9ed496/src/BLEMIDI_Settings.h#L7-L10

Midi lib has an [example](https://github.com/FortySevenEffects/arduino_midi_library/blob/2d64cc3c2ff85bbee654a7054e36c59694d8d8e4/examples/CustomBaudRate/CustomBaudRate.ino) when it uses a `struc` for override default settings, but I have tried something similar for this lib but unsuccessfully.

https://github.com/FortySevenEffects/arduino_midi_library/blob/2d64cc3c2ff85bbee654a7054e36c59694d8d8e4/examples/CustomBaudRate/CustomBaudRate.ino#L3-L7

https://github.com/FortySevenEffects/arduino_midi_library/blob/2d64cc3c2ff85bbee654a7054e36c59694d8d8e4/examples/CustomBaudRate/CustomBaudRate.ino#L14-L15

**Does exist any "clean way" to create custom settings in local code without modifying source lib files?**
Would you provide an example, please?

Thanks

**Comment by lathoub** [comment]
Hi @RobertoHE 
Yes, the BLE Settings in this lib are setup in the exact same way as in the MIDI lib:

https://github.com/lathoub/Arduino-BLE-MIDI/blob/8dee1f7d257aa71aeb0e956632da2dd2ae9ed496/src/BLEMIDI_Transport.h#L26

it just lacks an example for the moment. I'll see if i can create one tonight

**Comment by lathoub** [comment]
Here is how to override `BLEMIDI_NAMESPACE::DefaultSettings` , can you @RobertoHE  check before I commit?
(assuming this for an ESP32 using NimBLE, but is similar for the other hardware)

```cpp
#include <BLEMIDI_Transport.h>

struct CustomBufferSizeSettings : public BLEMIDI_NAMESPACE::DefaultSettings {
    static const size_t MaxBufferSize = 16; // was 64
};

#include <hardware/BLEMIDI_ESP32_NimBLE.h>
//#include <hardware/BLEMIDI_ESP32.h>
//#include <hardware/BLEMIDI_nRF52.h>
//#include <hardware/BLEMIDI_ArduinoBLE.h>

BLEMIDI_NAMESPACE::BLEMIDI_Transport<BLEMIDI_NAMESPACE::BLEMIDI_ESP32_NimBLE, CustomBufferSizeSettings> BLEMIDI("Esp32-NimBLE-MIDI"); \
MIDI_NAMESPACE::MidiInterface<BLEMIDI_NAMESPACE::BLEMIDI_Transport<BLEMIDI_NAMESPACE::BLEMIDI_ESP32_NimBLE>, BLEMIDI_NAMESPACE::MySettings> MIDI((BLEMIDI_NAMESPACE::BLEMIDI_Transport<BLEMIDI_NAMESPACE::BLEMIDI_ESP32_NimBLE> &)BLEMIDI);
```

I'm considering adding a `BLEMIDI_CREATE_CUSTOM_INSTANCE` macro for each of the hardware variants, similar to:

https://github.com/FortySevenEffects/arduino_midi_library/blob/2d64cc3c2ff85bbee654a7054e36c59694d8d8e4/src/serialMIDI.h#L128

**Comment by RobertoHE** [comment]
Hi @lathoub 

I have tested your example with Client hardware abstract definition of the template, but it didn't work for my case, but it compiled well.


I have seen the following behavior:

If I create a new config struct:
If `MaxBufferSize` is greater than 64, in execution time ESP32 core 1 is panic'ed because try to access a forbidden location.
If `MaxBufferSize` is lesser than 64, in execution time ESP32 is stuck at some point near scanning BLEs.

If I modify directly the struct of `BLEMIDI_Settings.h  (DefaultSettings)`, it works perfectly.

---
I am not sure, because working with templates is hard, but I think that hardware definitions of the template use the `default config` struct be "default".

https://github.com/lathoub/Arduino-BLE-MIDI/blob/8dee1f7d257aa71aeb0e956632da2dd2ae9ed496/src/BLEMIDI_Transport.h#L26

The definition of any of those templates use always the default value of `<... , class _Settings = DefaultSettings>` in each pointer to 'BLEMIDI_Transport'
https://github.com/lathoub/Arduino-BLE-MIDI/blob/8dee1f7d257aa71aeb0e956632da2dd2ae9ed496/src/hardware/BLEMIDI_Client_ESP32.h#L232

As I have said before, I am not sure, but it may be the cause that core 1 panic'ed because "default size (`DefaultSettings`)" that creates the buffer in the library is lesser than "local size" that is only used in the definition of main.c. I am not sure of that because any related to templates overcomes my knowledge ... and it seems that compiles XD.

Would you check if template definitions always use 'DefaultSettings' ?

Any suggestion is wellcomed.

Thank you

**Comment by lathoub** [comment]
> If I create a new config struct:
> If MaxBufferSize is greater than 64, in execution time ESP32 core 1 is panic'ed because try to access a forbidden location.
> If MaxBufferSize is lesser than 64, in execution time ESP32 is stuck at some point near scanning BLEs.
> 
> If I modify directly the struct of BLEMIDI_Settings.h (DefaultSettings), it works perfectly.

Ok, that indicates that the DefaultSetting are not overwritten (or related). Let see if I can get my hands on an ESP32

> working with templates is hard

nah, just a fancy replacing at compile time :-)

**Reply by lathoub** [reply]
hmm, harder than I thought... template argument also go into the hardware code. Still will take a bit longer

**Comment by lathoub** [comment]
Follow the work in [this branch](https://github.com/lathoub/Arduino-BLE-MIDI/tree/CustomSettings)

**Comment by lathoub** [comment]
> If I create a new config struct:
> If MaxBufferSize is greater than 64, in execution time ESP32 core 1 is panic'ed because try to access a forbidden location.
> If MaxBufferSize is lesser than 64, in execution time ESP32 is stuck at some point near scanning BLEs.

The buffer size is hardcoded here, and should be taken from the template. Already marked as a TODO (I now remember that I struggled with the templates before) - and explains the panic

https://github.com/lathoub/Arduino-BLE-MIDI/blob/2f9c684140affec506e759b9448918e32cc58461/src/hardware/BLEMIDI_ESP32_NimBLE.h#L125

**Comment by lathoub** [comment]
@RobertoHE  Please try with [this branch](https://github.com/lathoub/Arduino-BLE-MIDI/tree/CustomSettings) for the ESP32 (NimBLE) (I don't have a device here at hand, so no guarantees)

**Reply by RobertoHE** [reply]
It does not compile for me.

Error in define templates:

BLEMIDI_CREATE_INSTANCE("f2:c1:d9:aa:aa:aa", MIDIKEYTAR);   
or 
BLEMIDI_CREATE_CUSTOM_INSTANCE("f2:c1:d9:aa:aa:aa", MIDIKEYTAR, CustomSettings);
 
ERROR-> conflicting declaration 'midi::MidiInterface<bleMidi::BLEMIDI_Transport<bleMidi::BLEMIDI_Client_ESP32>, bleMidi::MySettings> MIDIKEYTAR'

**Reply by lathoub** [reply]
how about the new example? ([CustomerBufferSize](https://github.com/lathoub/Arduino-BLE-MIDI/blob/CustomSettings/examples/CustomerBufferSize/CustomerBufferSize.ino))

**Reply by RobertoHE** [reply]
The same error.

**Reply by RobertoHE** [reply]
Hi @lathoub 
Would you add BLEMIDI_Client_ESP32.h to the branch?

**Reply by lathoub** [reply]
Bizar, i have deleted all libs from the library folder and installed them fresh. MIDI_Library, NimBLE-Arduino and the `CustomSettings` Branch from this lib are the only ones in the lib folder now. Arduino IDE is 1.8.19 on MacOS. ESP32 board is `DOIT ESP32 DEVKIT v1`

My sketch header looks like:

```cpp
#include <BLEMIDI_Transport.h>

 struct CustomBufferSizeSettings : public BLEMIDI_NAMESPACE::DefaultSettings {
  static const size_t MaxBufferSize = 16;
};

#include <hardware/BLEMIDI_ESP32_NimBLE.h>
//#include <hardware/BLEMIDI_ESP32.h>
//#include <hardware/BLEMIDI_nRF52.h>
//#include <hardware/BLEMIDI_ArduinoBLE.h>

BLEMIDI_CREATE_CUSTOM_INSTANCE("Esp32-NimBLE-MIDI", MIDI, CustomBufferSizeSettings);
```

**Reply by lathoub** [reply]
> Hi @lathoub
> Would you add BLEMIDI_Client_ESP32.h to the branch?

It is already - which signals to me that your lib is somewhat out of whack

**Reply by RobertoHE** [reply]
BLEMIDI_Client_ESP32.h file hasn't any new commits in that branch.

Would you modify this file to adapt to the new style of templates?

This evening (maybe) I will try to code it, but I would like to start from a file that implements the new style templates because I am not familiarized with templates inside templates inside other templates. Yesterday tried to code it but I couldn't do it successfully. Today I will try again.

**Reply by lathoub** [reply]
> Would you modify this file to adapt to the new style of templates?

done

**Reply by RobertoHE** [reply]
Pull request done.

Let me some days to test it properly and change the actual settings based on hardcoded defines by an elegant config struct.

**Reply by lathoub** [reply]
I did some additional fixes, make sure you use always fetch the latest
